pytissueoptics/rayscattering/opencl/CLPhotons.py

```diff
@@ -0,0 +1,168 @@
+                 ("material_id", cl.cltypes.uint)])
+            name = "photonStruct"
+            photonStruct, c_decl_photon = cl.tools.match_dtype_to_c_struct(self._device, name, photonStruct)
+            photon_dtype = cl.tools.get_or_register_dtype(name, photonStruct)
+            return photon_dtype, c_decl_photon
+
+        def makeMaterialType():
+            materialStruct = np.dtype(
+                [("mu_s", cl.cltypes.float),
+                 ("mu_a", cl.cltypes.float),
+                 ("mu_t", cl.cltypes.float),
+                 ("g", cl.cltypes.float),
+                 ("n", cl.cltypes.float),
+                 ("albedo", cl.cltypes.float),
+                 ("material_id", cl.cltypes.uint)])
+            name = "materialStruct"
+            materialStruct, c_decl_mat = cl.tools.match_dtype_to_c_struct(self._device, name, materialStruct)
+            material_dtype = cl.tools.get_or_register_dtype(name, materialStruct)
+            return material_dtype, c_decl_mat
+
+        def makeLoggerType():
+            loggerStruct = np.dtype(
+                [("delta_weight", cl.cltypes.float),
+                 ("x", cl.cltypes.float),
+                 ("y", cl.cltypes.float),
+                 ("z", cl.cltypes.float)])
+            name = "loggerStruct"
+            loggerStruct, c_decl_logger = cl.tools.match_dtype_to_c_struct(self._device, name, loggerStruct)
+            logger_dtype = cl.tools.get_or_register_dtype(name, loggerStruct)
+            return logger_dtype, c_decl_logger
+
+        self._photon_dtype, self._c_decl_photon = makePhotonType()
+        self._material_dtype, self._c_decl_mat = makeMaterialType()
+        self._logger_dtype, self._c_decl_logger = makeLoggerType()
+
+    def _makeBuffers(self):
+        self._makePhotonsBuffer()
+        self._makeMaterialsBuffer()
+        self._makeLoggerBuffer()
+        self._makeRandomBuffer()
+
+    def _makePhotonsBuffer(self):
+        self._HOST_photons = np.empty(self._source.getPhotonCount(), dtype=self._photon_dtype)
+
+        self._DEVICE_photons = cl.Buffer(self._context, cl.mem_flags.READ_WRITE | cl.mem_flags.COPY_HOST_PTR,
+                                         hostbuf=self._HOST_photons)
+
+    def _makeRandomBuffer(self):
+        self._HOST_randomSeed = np.random.randint(low=0, high=2 ** 32 - 1, size=self._source.getPhotonCount(),
+                                                  dtype=cl.cltypes.uint)
+        self._DEVICE_randomSeed = cl.Buffer(self._context, cl.mem_flags.READ_WRITE | cl.mem_flags.COPY_HOST_PTR,
+                                            hostbuf=self._HOST_randomSeed)
+        self._HOST_randomFloat = np.empty(self._source.getPhotonCount(), dtype=cl.cltypes.float)
+        self._DEVICE_randomFloat = cl.Buffer(self._context, cl.mem_flags.READ_WRITE | cl.mem_flags.COPY_HOST_PTR,
+                                             hostbuf=self._HOST_randomFloat)
+
+    def _makeMaterialsBuffer(self):
```

Comment:
(Following the comment regarding duplicated code in the test file) There is a lot of CL buffer, CL types and other CL sauce that we could _consider_ moving out of this Photons object.

Reply:
Yes I agree

Reply:
CLBuffer, I'm not sure moving it out will be easy, I'll think about it.

Reply:
Yeah I don't think it is necessary or useful for now. It'll add a lot of blabla.  Each buffer takes specific things, need access to the context, and we want photon to have access to the HOST to access the data and DEVICE to call kernels